src/main/java/org/ngafid/common/TimeUtils.java

```diff
@@ -259,4 +282,171 @@ public static Date addMilliseconds(Date date, Integer milliseconds) {
+            ZoneId zoneId = ZoneId.of(zoneIdStr);
+
+            // Convert the UTC datetime to the local time in the specified zone.
+            ZonedDateTime localZonedDateTime = utcDateTime.atZone(ZoneOffset.UTC).withZoneSameInstant(zoneId);
+
+            // Format the local date, local time, and UTC offset.
+            localDates.add(localZonedDateTime.toLocalDate().format(DateTimeFormatter.ofPattern("yyyy-MM-dd")));
+            localTimes.add(localZonedDateTime.toLocalTime().format(DateTimeFormatter.ofPattern("HH:mm:ss")));
+            utcOffsets.add(localZonedDateTime.getOffset().getId());
+        }
+
+        return new LocalDateTimeResult(localDates, localTimes, utcOffsets);
+    }
+
+
+    /**
+     * Will remove this method after approval.
+     * Calculates local date, local time, and UTC offset for each entry based on given UTC dates, times, latitudes, and longitudes.
+     * @param map          map object (created once per upload for efficiency.
+     * @param utcDates     list of UTC dates as strings (e.g., "yyyy-MM-dd")
+     * @param utcTimes     list of UTC times as strings (e.g., "HH:mm:ss")
+     * @param latitudes    list of latitude strings
+     * @param longitudes   list of longitude strings
+     * @return a LocalDateTimeResult containing lists of local dates, times, and UTC offsets.
+     */
+    public static LocalDateTimeResult calculateLocalDateTime(TimeZoneMap map,
+                                                             ArrayList<String> utcDates, ArrayList<String> utcTimes,
+                                                             ArrayList<String> latitudes, ArrayList<String> longitudes) throws Exception {
+
+        // Validate that all lists are of the same size.
+        if (utcDates.size() != utcTimes.size() || utcDates.size() != latitudes.size() || utcDates.size() != longitudes.size()) {
+            throw new IllegalArgumentException("All input lists must have the same size.");
+        }
+
+        // Prepare lists for the results.
+        ArrayList<String> localDates = new ArrayList<>(utcDates.size());
+        ArrayList<String> localTimes = new ArrayList<>(utcDates.size());
+        ArrayList<String> utcOffsets = new ArrayList<>(utcDates.size());
+
+        // Iterate over each index and calculate the corresponding local date, time, and offset.
+        for (int i = 0; i < utcDates.size(); i++) {
+            String dateTimeString = utcDates.get(i) + " " + utcTimes.get(i);
+            LocalDateTime utcDateTime = null;
+
+            // Try parsing using a formatter.
+            for (DateTimeFormatter formatter : DATE_FORMATTERS) {
+                try {
+                    utcDateTime = LocalDateTime.parse(dateTimeString, formatter);
+                    break;
+                } catch (DateTimeParseException e) {
+                    // Continue trying other formats
+                }
+            }
```

Comment:
Determine the date format once and use it at every time step. Move this out of the loop.